skynet/src/main/java/org/example/service/Faction.java

```diff
@@ -0,0 +1,133 @@
+        this.title = title;
+        this.storage = storage;
+        this.DAYS = DAYS;
+        this.MAX_DAY_CAPACITY = MAX_CAPACITY;
+        this.inventory = new EnumMap<>(Part.class);
+
+        for (Part part : Part.values()) {
+            inventory.put(part, 0);
+        }
+    }
+
+    /**
+     * The main execution method for the faction thread.
+     * Runs the building cycle for the specified number of days.
+     * Handles interruptions gracefully and logs appropriate messages.
+     */
+    @Override
+    public void run() {
+        logger.info("Faction  " + title + " is starting");
+        try {
+            executeBuildingCycle();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            logger.severe("Faction " + title + " was interrupted " + e);
+        } finally {
+            logger.fine("Faction " + title + " finished");
+        }
+    }
+
+    /**
+     * Attempts to build as many robots as possible from available parts.
+     * Continues building until parts remain for another robot.
+     */
+    private void buildRobots() {
+        while (canBuild()) {
+            consumeParts();
+            robotsBuilt++;
+        }
+    }
+
+    /**
+     * Checks if there are enough parts to build a complete robot.
+     *
+     * @return true if all required parts are available in enough quantities, false otherwise
+     */
+    private boolean canBuild() {
+        return inventory.getOrDefault(Part.HEAD, 0) >= REQUIRED_HEAD &&
+                inventory.getOrDefault(Part.TORSO, 0) >= REQUIRED_TORSO &&
+                inventory.getOrDefault(Part.HAND, 0) >= REQUIRED_HANDS &&
+                inventory.getOrDefault(Part.FEET, 0) >= REQUIRED_FEET;
+    }
+
+    /**
+     * Consumes the required parts from inventory to build one robot.
+     * Reduces the inventory counts for each part type by the required amount.
+     */
+    private void consumeParts() {
+        inventory.merge(Part.HEAD, -REQUIRED_HEAD, Integer::sum);
+        inventory.merge(Part.TORSO, -REQUIRED_TORSO, Integer::sum);
+        inventory.merge(Part.HAND, -REQUIRED_HANDS, Integer::sum);
+        inventory.merge(Part.FEET, -REQUIRED_FEET, Integer::sum);
+    }
+
+    /**
+     * Executes the complete building cycle for all days.
+     * Each day consists of collecting parts and attempting to build robots.
+     *
+     * @throws InterruptedException if the thread is interrupted
+     */
+    private void executeBuildingCycle() throws InterruptedException {
+        for (int day = 1; day <= DAYS; day++) {
+            collectPartsForDay();
+            buildRobots();
+            Thread.sleep(50);
+        }
+    }
+
+    /**
+     * Collects parts from the shared storage for the current day.
+     * Attempts to collect up to the maximum daily capacity, but stops early if storage is empty.
+     */
+    private void collectPartsForDay() {
```

Comment:
poll()  is non-blocking. This doesn't properly represent the requirement that factions collect parts "at night" - there's no coordination between day/night cycles.